### PR TITLE
ItemBoxAPI2

### PR DESCRIPTION
**Pull request by windingwind**

### CODE REVIEW

**Comment by dstillman** [comment]
Can you rebase this on current main?

**Comment by windingwind** [comment]
rebased

File: chrome/content/zotero/xpcom/itemBoxManager.js [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+	 *   editable: true,
+	 *   multiline: true,
+	 *   index: 1,
+	 *   dataProvider: (item, dataKey) => item.getField('abstractNote').split('').reverse().join(''),
+	 *   dataSetter: (item, dataKey, value) => Zotero.debug(`Item ${item.getField("title")}'s reversed abstract is ${value} now.`),
+	 *   expandStateGetter: (item, dataKey) => Zotero.Prefs.get("extensions.make-it-red.reversedAbstractExpanded", true),
+	 *   expandStateSetter: (item, dataKey, expanded) => Zotero.Prefs.set("extensions.make-it-red.reversedAbstractExpanded", expanded, true)
+	 * });
+	 * ```
+	 */
```
Comment (windingwind) [inline chrome/content/zotero/xpcom/itemBoxManager.js:91]: Examples are here

File: chrome/content/zotero/elements/itemBox.js [hunk 2]
```diff
@@ -552,45 +584,57 @@
...
 				let valueElement = this.createValueElement(
 					val, fieldName, tabindex
 				);
+
+				if (!valueElement) {
+					continue;
+				}
 				
 				var prefix = '';
 				// Add '(...)' before 'Abstract' for collapsed abstracts
```
[the remark this answers is not in the document]
Comment (windingwind) [reply chrome/content/zotero/elements/itemBox.js]: done

File: chrome/content/zotero/elements/itemBox.js [hunk 3]
```diff
@@ -1056,7 +1102,7 @@
 			th.setAttribute("onclick", "this.nextSibling.firstChild.blur()");
```
[the remark this answers is not in the document]
Comment (windingwind) [reply chrome/content/zotero/elements/itemBox.js]: done

File: chrome/content/zotero/xpcom/itemBoxManager.js [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+ * - Rows with lower indices are displayed first.
+ * - Rows with the same index are displayed in unknown order.
+ * - Index starts from `1`, after item type, title, and creators.
+ * @property {boolean} [editable=false] - Whether the row is editable.
+ * @property {boolean} [multiline=false] - Whether the row is multiline.
+ * @property {(item: Zotero.Item, dataKey: string) => string} dataProvider
+ * - Custom data provider that is called when rendering rows.
+ * @property {(item: Zotero.Item, dataKey: string, value: string) => void} dataSetter
+ * - Custom data setter that is called when editing rows.
+ * @property {(item: Zotero.Item, dataKey: string) => boolean} [expandStateGetter]
```
[the remark this answers is not in the document]
Comment (windingwind) [reply chrome/content/zotero/xpcom/itemBoxManager.js]: done

File: chrome/content/zotero/xpcom/itemBoxManager.js [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+ * @property {string} label - Label for the row to be displayed in the item box.
+ * @property {string} pluginID - Set plugin ID to auto remove row when plugin is removed.
+ * @property {number} [index] - Index of the row.
+ * - If not set, the row is appended to the end.
+ * - Rows with lower indices are displayed first.
+ * - Rows with the same index are displayed in unknown order.
+ * - Index starts from `1`, after item type, title, and creators.
+ * @property {boolean} [editable=false] - Whether the row is editable.
+ * @property {boolean} [multiline=false] - Whether the row is multiline.
+ * @property {(item: Zotero.Item, dataKey: string) => string} dataProvider
```
[the remark this answers is not in the document]
Comment (windingwind) [reply chrome/content/zotero/xpcom/itemBoxManager.js]: done

**Comment by timokau** [comment]
For reference: I believe this is the PR hinted at [in the zotero-dev group](https://groups.google.com/g/zotero-dev/c/dzBI00ptxGw/m/EA7LS6JyAAAJ) which removes the final remaining blocker for better bibtex Zotero 7 support (https://github.com/retorquere/zotero-better-bibtex/issues/2522).

**Pull request closed without merging**